File: leetcode/0091-numDecodings.py
class Solution:
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        str_len = len(s)
        dp = [0 for _ in range(str_len+1)]
        dp[str_len] = 1
        dp[str_len-1] = 0 if s[str_len-1:str_len] == "0" else 1
        for i in range(str_len-2, -1, -1):
            if s[i:i+1] == "0":
                continue
            code = int(s[i:i+2])
            dp[i] = dp[i+1] + dp[i+2] if code <= 26 else dp[i+1]
            if dp[i] == 0:
                break
        return dp[0]

File: leetcode/0091-numDecodings.java
class Solution {
    public int numDecodings(String s) {
        if (s == null || s.length() == 0)
            return 0;
        int len = s.length();
        int[] dp = new int[len+2];
        dp[len] = dp[len+1] = 1;
        for (int i = len-1; i >= 0; i--) {
            int oneBitVal = Integer.valueOf(s.substring(i, i+1));
            if (oneBitVal == 0) {
                dp[i] = 0;
                continue;
            } else {
                dp[i] += dp[i+1];
            }
            if (i+2 <= len) {
                int twoBitVal = Integer.valueOf(s.substring(i, i+2));
                if (twoBitVal >= 1 && twoBitVal <= 26)
                    dp[i] += dp[i+2];
            }
        }
        return dp[0];
    }
}
